{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "553a6db9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from fun import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4fb932c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "db_filename = '/home/msarrias/dbs/homo_1.db'\n",
    "db = gffutils.FeatureDB(db_filename, keep_order=True)\n",
    "fasta_GRCh38_seq = SeqIO.parse(open('/home/msarrias/data/Homo_sapiens.GRCh38.dna.chromosome.1.fa'),'fasta')\n",
    "fasta_GRCh38_seq = { fasta.id : {'+' : str(fasta.seq),\n",
    "                                 '-': str(fasta.seq.reverse_complement())} for fasta in fasta_GRCh38_seq}\n",
    "\n",
    "gene_hierarchy_dict = generate_gene_hierarchy_dict(db)\n",
    "gene_transcript_dict, gene_hierarchy_dict_with_ce = get_gene_hierarchy_dict_with_coding_exons(gene_hierarchy_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "18e23819",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(10)\n",
    "max_n_exons = 10\n",
    "sample_genes, simulated_genes, sample_transcripts_dict = {}, {}, {}\n",
    "copy_gene_hierarchy_dict_with_ce = copy.deepcopy(gene_hierarchy_dict_with_ce)\n",
    "settings_list = sorted([(n_exons, np.random.randint(2, n_exons-1))\n",
    "                        for n_exons in random.sample(range(4,max_n_exons+1), 7)],\n",
    "                       key = lambda item:item[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "321ae8a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "(4, 2) before\n",
      "gene:ENSG00000067606 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162592 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000173662 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000116273 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000041988 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000171735 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162444 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000116663 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000204518 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000188984 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000065526 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000159339 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000159339 split intron ends with:  AG     starts with:  TC\n",
      "gene:ENSG00000270136 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000270136 split intron ends with:  AG     starts with:  CT\n",
      "gene:ENSG00000127472 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000158825 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162551 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000070831 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117602 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000131914 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117682 split intron ends with:  AG     starts with:  GT\n",
      "    \n",
      "True\n",
      "(4, 2) after\n",
      "gene:ENSG00000067606 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162592 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000173662 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000116273 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000041988 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000171735 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162444 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000116663 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000204518 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000188984 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000065526 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000159339 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000159339 split intron ends with:  AG     starts with:  TC\n",
      "gene:ENSG00000270136 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000270136 split intron ends with:  AG     starts with:  CT\n",
      "gene:ENSG00000127472 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000158825 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162551 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000070831 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117602 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000131914 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117682 split intron ends with:  AG     starts with:  GT\n",
      "    \n",
      "True\n",
      "(5, 2) before\n",
      "gene:ENSG00000215915 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000215915 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157916 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157870 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000196581 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000041988 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000049239 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000171819 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000204624 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000132879 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000028137 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162493 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142634 split intron ends with:  GG     starts with:  GT\n",
      "gene:ENSG00000142634 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162461 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000158786 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000187942 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000219073 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000219073 split intron ends with:  AG     starts with:  GA\n",
      "gene:ENSG00000011009 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000001461 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117682 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000198492 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000198492 split intron ends with:  AG     starts with:  AA\n",
      "    \n",
      "True\n",
      "(5, 2) after\n",
      "gene:ENSG00000215915 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000215915 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157916 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157870 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000196581 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000041988 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000049239 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000171819 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000204624 split intron ends with:  AG     starts with:  GC\n",
      "gene:ENSG00000132879 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000028137 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162493 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142634 split intron ends with:  GG     starts with:  GT\n",
      "gene:ENSG00000142634 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162461 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000158786 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000187942 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000219073 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000219073 split intron ends with:  AG     starts with:  GA\n",
      "gene:ENSG00000011009 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000001461 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117682 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000198492 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000198492 split intron ends with:  AG     starts with:  AA\n",
      "    \n",
      "True\n",
      "(6, 2) before\n",
      "gene:ENSG00000169962 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000169885 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157916 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157873 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157873 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000116288 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142655 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162490 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162460 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000169504 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000183726 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000198830 split intron ends with:  CC     starts with:  GC\n",
      "gene:ENSG00000198830 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000121766 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000126070 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000116863 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000163879 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000084072 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000084072 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000084072 split intron ends with:  AG     starts with:  AA\n",
      "gene:ENSG00000117411 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117411 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117411 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117411 split intron ends with:  AG     starts with:  CA\n",
      "gene:ENSG00000142937 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000222009 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162368 split intron ends with:  AG     starts with:  GT\n",
      "    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "(6, 2) after\n",
      "gene:ENSG00000169962 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000169885 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157916 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157873 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157873 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000116288 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142655 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162490 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162460 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000169504 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000183726 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000198830 split intron ends with:  CC     starts with:  GC\n",
      "gene:ENSG00000198830 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000121766 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000126070 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000116863 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000163879 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000084072 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000084072 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000084072 split intron ends with:  AG     starts with:  AA\n",
      "gene:ENSG00000117411 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117411 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117411 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117411 split intron ends with:  AG     starts with:  CA\n",
      "gene:ENSG00000142937 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000222009 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162368 split intron ends with:  AG     starts with:  GT\n",
      "    \n",
      "True\n",
      "(7, 3) before\n",
      "gene:ENSG00000162572 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157933 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157870 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000169598 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000171612 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000065526 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000204178 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117751 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000180198 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000121764 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000025800 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162521 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162522 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000121903 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142698 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000020129 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000116819 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000214193 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000183682 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000168389 split intron ends with:  AG     starts with:  GT\n",
      "    \n",
      "True\n",
      "(7, 3) after\n",
      "gene:ENSG00000162572 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157933 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000157870 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000169598 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000171612 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000065526 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000204178 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117751 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000180198 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000121764 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000025800 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162521 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162522 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000121903 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142698 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000020129 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000116819 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000214193 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000183682 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000168389 split intron ends with:  AG     starts with:  GT\n",
      "    \n",
      "True\n",
      "(8, 6) before\n",
      "gene:ENSG00000169972 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162426 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000131686 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000131686 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000131686 split intron ends with:  AG     starts with:  CC\n",
      "gene:ENSG00000188807 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000130939 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142615 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142615 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142615 split intron ends with:  AG     starts with:  CC\n",
      "gene:ENSG00000215704 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000215704 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000215704 split intron ends with:  AG     starts with:  CC\n",
      "gene:ENSG00000053372 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000169914 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000158828 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000219073 split intron ends with:  TC     starts with:  GT\n",
      "gene:ENSG00000219073 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000219073 split intron ends with:  AG     starts with:  AA\n",
      "gene:ENSG00000142789 split intron ends with:  TC     starts with:  GT\n",
      "gene:ENSG00000142789 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142789 split intron ends with:  AG     starts with:  AA\n",
      "gene:ENSG00000188782 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000009780 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000160094 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000183520 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117000 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000222009 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000132773 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000088035 split intron ends with:  AG     starts with:  GT\n",
      "    \n",
      "True\n",
      "(8, 6) after\n",
      "gene:ENSG00000169972 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000162426 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000131686 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000131686 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000131686 split intron ends with:  AG     starts with:  CC\n",
      "gene:ENSG00000188807 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000130939 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142615 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142615 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142615 split intron ends with:  AG     starts with:  CC\n",
      "gene:ENSG00000215704 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000215704 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000215704 split intron ends with:  AG     starts with:  CC\n",
      "gene:ENSG00000053372 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000169914 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000158828 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000219073 split intron ends with:  TC     starts with:  GT\n",
      "gene:ENSG00000219073 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000219073 split intron ends with:  AG     starts with:  AA\n",
      "gene:ENSG00000142789 split intron ends with:  TC     starts with:  GT\n",
      "gene:ENSG00000142789 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000142789 split intron ends with:  AG     starts with:  AA\n",
      "gene:ENSG00000188782 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000009780 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000160094 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000183520 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000117000 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000222009 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000132773 split intron ends with:  AG     starts with:  GT\n",
      "gene:ENSG00000088035 split intron ends with:  AG     starts with:  GT\n",
      "    \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "for setting in settings_list:\n",
    "    n_exons, exon_dup = setting\n",
    "    sample_genes[setting] = {}\n",
    "    sample_transcripts_dict[setting] = {}\n",
    "    for gene_id, gene_hierarchy_dict_ in copy_gene_hierarchy_dict_with_ce.items():\n",
    "        if db[gene_id].strand == '+' and len(gene_hierarchy_dict_) > 0:\n",
    "            transcript = random.choice(list(gene_hierarchy_dict_.keys()))\n",
    "            transcript_temp = {}\n",
    "            trapt_hier_dict = copy.deepcopy(gene_hierarchy_dict_[transcript])\n",
    "            n_coding_exons = [i['type'] for i in trapt_hier_dict.values() if i['type'] == 'coding_exon']\n",
    "            if len(n_coding_exons) == n_exons:\n",
    "                sample_genes[setting][gene_id] = copy.deepcopy(trapt_hier_dict)\n",
    "                sample_transcripts_dict[setting][gene_id] = transcript\n",
    "            if len(sample_genes[setting].keys()) > 19:\n",
    "                break\n",
    "    simulated_genes[setting] = {}\n",
    "    for pos in ['before', 'after']:\n",
    "        sim_genes_sample = simulate_genes_with_coding_exons_duplications(exon_dup, \n",
    "                                                                          pos, \n",
    "                                                                          sample_genes[setting],\n",
    "                                                                          gene_hierarchy_dict_with_ce,\n",
    "                                                                          fasta_GRCh38_seq['1']['+'],\n",
    "                                                                          db)\n",
    "        sorted_gene_sample = {gene[0]: copy.deepcopy(sim_genes_sample[gene[0]]) \n",
    "                              for gene in sorted([(key, value['coord']) \n",
    "                                                  for key, value in sim_genes_sample.items()],\n",
    "                                                 key = lambda item: (item[1].lower, item[1].upper))\n",
    "                     }\n",
    "        simulated_genes[setting][pos] = sorted_gene_sample\n",
    "        new_seq, len_dup = generate_new_genome_seq(sorted_gene_sample,\n",
    "                                                   gene_hierarchy_dict_with_ce,\n",
    "                                                   fasta_GRCh38_seq['1']['+'], db)\n",
    "        print(len(new_seq) - len_dup == len(fasta_GRCh38_seq['1']['+']))\n",
    "        coding_exons_interv = {gene_id :\n",
    "                               {i:j for i, j in zip(transcr_dict['features_order'],\n",
    "                                                    transcr_dict['features_intervals']) if 'exon' in i} \n",
    "                               for gene_id, transcr_dict in sorted_gene_sample.items() \n",
    "                              }\n",
    "        sim_genes_interv_in_chr = get_simulated_new_intervals(fasta_GRCh38_seq['1']['+'],\n",
    "                                                              new_seq,\n",
    "                                                              sorted_gene_sample,\n",
    "                                                              coding_exons_interv)\n",
    "        print(setting, pos)\n",
    "        for gene_id, gene_dict in sim_genes_interv_in_chr.items():\n",
    "            for id_, annot in gene_dict['coding_exons_dups'].items():\n",
    "                if len(annot) == 2:\n",
    "                    s, e = annot[0]\n",
    "                    print(gene_id, 'split intron ends with: ', new_seq[(s-2):(s)],\n",
    "                          '   ', 'starts with: ', new_seq[e:(e+2)])\n",
    "        print('    ')            \n",
    "        with open('../files/GRCh38_dup_' + str(n_exons)+ '_' + str(exon_dup) + '_' + pos + '.pkl', 'wb') as handle:\n",
    "            pickle.dump(sim_genes_interv_in_chr, handle)\n",
    "\n",
    "        record = SeqRecord(\n",
    "            Seq(new_seq),\n",
    "            id = '+')\n",
    "        with open(\"GRCh38_dup_\" + str(n_exons)+ '_' + str(exon_dup) + '_' + pos + \".fa\", \"w\") as handle:\n",
    "            SeqIO.write(record, handle, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45cb60f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_protein_seqs = {}\n",
    "fasta_sequences = SeqIO.parse(open(\"fasta_files/Homo_sapiens.GRCh38.pep.all.fa\"),'fasta')\n",
    "for fasta in fasta_sequences:\n",
    "    gene = [i for i in fasta.description.rsplit(' ') if 'gene:' in i]\n",
    "    if gene:\n",
    "        if gene[0] not in human_protein_seqs:\n",
    "            human_protein_seqs[gene[0]] = {}\n",
    "        human_protein_seqs[gene[0]][fasta.id] = str(fasta.seq)\n",
    "        \n",
    "for setting, setting_dict in sample_genes.items():\n",
    "    genes_list = list(setting_dict.keys())\n",
    "    proteins_simulated_genes = {}\n",
    "    protein_gene_dict = {}\n",
    "    with open(\"fasta_files/prot_seq_\" +str(setting[0]) + '_' + str(setting[1]) + \".fa\", \"w\") as handle:\n",
    "        for gene in genes_list:\n",
    "            gene_id_prot_dict = [i for i in list(human_protein_seqs.keys()) if gene in i]\n",
    "            if gene_id_prot_dict:\n",
    "                temp = copy.deepcopy(human_protein_seqs[gene_id_prot_dict[0]])\n",
    "                prot_random = random.choice(list(temp.keys()))\n",
    "                protein_gene_dict[prot_random.split('.')[0]] = gene_id_prot_dict[0].split('.')[0]\n",
    "                record = SeqRecord(Seq(temp[prot_random]),\n",
    "                                   id = str(prot_random),\n",
    "                                   description= gene_id_prot_dict[0])\n",
    "                SeqIO.write(record, handle, \"fasta\")\n",
    "                proteins_simulated_genes[gene_id_prot_dict[0]] = temp[prot_random]\n",
    "    with open(\"../files/protein_gene_corresp_\" + str(setting[0]) + '_' + str(setting[1]) + \".pkl\", 'wb') as handle:\n",
    "        pickle.dump(protein_gene_dict, handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30f4dce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../files/transcripts_dict_.pkl\", 'wb') as handle:\n",
    "    pickle.dump({'simulated_genes':simulated_genes,\n",
    "                 'sample_genes':sample_genes,\n",
    "                 'sample_transcripts_dict':sample_transcripts_dict},\n",
    "                handle)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4f45857",
   "metadata": {},
   "source": [
    "## Exonerate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "071f7752",
   "metadata": {},
   "outputs": [],
   "source": [
    "for setting in settings_list[2:]:\n",
    "    queries = \"fasta_files/prot_seq_\" +str(setting[0]) + '_' + str(setting[1]) + \".fa\"\n",
    "    for pos in ['before', 'after']:\n",
    "        hit = \"fasta_files/GRCh38_dup_\" + str(setting[0]) + '_' + str(setting[1]) +'_' + pos +  \".fa\"\n",
    "        out_file = 'Exonerate/exonerate_' + str(setting[0]) + '_' + str(setting[1]) +'_' + pos + '.txt'\n",
    "        f = open(out_file, \"w\")\n",
    "        subprocess.call([\"exonerate\", \"--query\", queries, '--target',  hit, '--model', 'protein2genome'], stdout=f)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "848c0c69",
   "metadata": {},
   "source": [
    "## Miniprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e4f00822",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::mp_ntseq_read@1.254*1.02] read 248883353 bases in 1 contigs\n",
      "[M::mp_idx_build@1.254*1.02] 1944402 blocks\n",
      "[M::mp_idx_build@17.280*0.99] collected syncmers\n",
      "[M::mp_idx_build@18.023*0.99] 70569585 kmer-block pairs\n",
      "[M::mp_idx_print_stat] 1101931 distinct k-mers; mean occ of infrequent k-mers: 64.04; 0 frequent k-mers accounting for 0 occurrences\n",
      "[M::worker_pipeline::19.029*0.99] mapped 40 sequences\n",
      "[M::main] Version: 0.1-r97\n",
      "[M::main] CMD: /domus/h1/msarrias/bin/miniprot/miniprot -t8 fasta_files/GRCh38_with_ce_dup_6_2_before.fa fasta_files/prot_seq6_2.fa\n",
      "[M::main] Real time: 19.075 sec; CPU: 18.908 sec; Peak RSS: 1.599 GB\n",
      "[M::mp_ntseq_read@1.255*1.02] read 248883353 bases in 1 contigs\n",
      "[M::mp_idx_build@1.255*1.02] 1944402 blocks\n",
      "[M::mp_idx_build@17.124*0.99] collected syncmers\n",
      "[M::mp_idx_build@17.827*0.99] 70569607 kmer-block pairs\n",
      "[M::mp_idx_print_stat] 1101931 distinct k-mers; mean occ of infrequent k-mers: 64.04; 0 frequent k-mers accounting for 0 occurrences\n",
      "[M::worker_pipeline::18.818*0.99] mapped 40 sequences\n",
      "[M::main] Version: 0.1-r97\n",
      "[M::main] CMD: /domus/h1/msarrias/bin/miniprot/miniprot -t8 fasta_files/GRCh38_with_ce_dup_6_2_after.fa fasta_files/prot_seq6_2.fa\n",
      "[M::main] Real time: 18.868 sec; CPU: 18.707 sec; Peak RSS: 1.599 GB\n",
      "[M::mp_ntseq_read@1.247*1.02] read 247784295 bases in 1 contigs\n",
      "[M::mp_idx_build@1.247*1.02] 1935816 blocks\n",
      "[M::mp_idx_build@17.115*0.99] collected syncmers\n",
      "[M::mp_idx_build@17.808*0.99] 70218179 kmer-block pairs\n",
      "[M::mp_idx_print_stat] 1101669 distinct k-mers; mean occ of infrequent k-mers: 63.74; 0 frequent k-mers accounting for 0 occurrences\n",
      "[M::worker_pipeline::18.800*0.99] mapped 40 sequences\n",
      "[M::main] Version: 0.1-r97\n",
      "[M::main] CMD: /domus/h1/msarrias/bin/miniprot/miniprot -t8 fasta_files/GRCh38_with_ce_dup_7_5_before.fa fasta_files/prot_seq7_5.fa\n",
      "[M::main] Real time: 18.845 sec; CPU: 18.665 sec; Peak RSS: 1.599 GB\n",
      "[M::mp_ntseq_read@1.248*1.02] read 247784295 bases in 1 contigs\n",
      "[M::mp_idx_build@1.249*1.02] 1935816 blocks\n",
      "[M::mp_idx_build@17.127*0.99] collected syncmers\n",
      "[M::mp_idx_build@17.836*0.99] 70218243 kmer-block pairs\n",
      "[M::mp_idx_print_stat] 1101668 distinct k-mers; mean occ of infrequent k-mers: 63.74; 0 frequent k-mers accounting for 0 occurrences\n",
      "[M::worker_pipeline::18.829*0.99] mapped 40 sequences\n",
      "[M::main] Version: 0.1-r97\n",
      "[M::main] CMD: /domus/h1/msarrias/bin/miniprot/miniprot -t8 fasta_files/GRCh38_with_ce_dup_7_5_after.fa fasta_files/prot_seq7_5.fa\n",
      "[M::main] Real time: 18.876 sec; CPU: 18.714 sec; Peak RSS: 1.599 GB\n",
      "[M::mp_ntseq_read@1.252*1.02] read 248842685 bases in 1 contigs\n",
      "[M::mp_idx_build@1.252*1.02] 1944084 blocks\n",
      "[M::mp_idx_build@17.220*0.99] collected syncmers\n",
      "[M::mp_idx_build@17.919*0.99] 70557943 kmer-block pairs\n",
      "[M::mp_idx_print_stat] 1101922 distinct k-mers; mean occ of infrequent k-mers: 64.03; 0 frequent k-mers accounting for 0 occurrences\n",
      "[M::worker_pipeline::18.909*0.99] mapped 40 sequences\n",
      "[M::main] Version: 0.1-r97\n",
      "[M::main] CMD: /domus/h1/msarrias/bin/miniprot/miniprot -t8 fasta_files/GRCh38_with_ce_dup_8_3_before.fa fasta_files/prot_seq8_3.fa\n",
      "[M::main] Real time: 18.954 sec; CPU: 18.788 sec; Peak RSS: 1.599 GB\n",
      "[M::mp_ntseq_read@1.258*1.02] read 248842685 bases in 1 contigs\n",
      "[M::mp_idx_build@1.259*1.02] 1944084 blocks\n",
      "[M::mp_idx_build@17.145*0.99] collected syncmers\n",
      "[M::mp_idx_build@17.852*0.99] 70557961 kmer-block pairs\n",
      "[M::mp_idx_print_stat] 1101922 distinct k-mers; mean occ of infrequent k-mers: 64.03; 0 frequent k-mers accounting for 0 occurrences\n",
      "[M::worker_pipeline::18.842*0.99] mapped 40 sequences\n",
      "[M::main] Version: 0.1-r97\n",
      "[M::main] CMD: /domus/h1/msarrias/bin/miniprot/miniprot -t8 fasta_files/GRCh38_with_ce_dup_8_3_after.fa fasta_files/prot_seq8_3.fa\n",
      "[M::main] Real time: 18.888 sec; CPU: 18.715 sec; Peak RSS: 1.599 GB\n",
      "[M::mp_ntseq_read@1.252*1.02] read 248741500 bases in 1 contigs\n",
      "[M::mp_idx_build@1.253*1.02] 1943294 blocks\n",
      "[M::mp_idx_build@17.137*0.99] collected syncmers\n",
      "[M::mp_idx_build@17.837*0.99] 70529535 kmer-block pairs\n",
      "[M::mp_idx_print_stat] 1101917 distinct k-mers; mean occ of infrequent k-mers: 64.01; 0 frequent k-mers accounting for 0 occurrences\n",
      "[M::worker_pipeline::19.043*0.99] mapped 40 sequences\n",
      "[M::main] Version: 0.1-r97\n",
      "[M::main] CMD: /domus/h1/msarrias/bin/miniprot/miniprot -t8 fasta_files/GRCh38_with_ce_dup_9_2_before.fa fasta_files/prot_seq9_2.fa\n",
      "[M::main] Real time: 19.114 sec; CPU: 18.948 sec; Peak RSS: 1.599 GB\n",
      "[M::mp_ntseq_read@1.252*1.02] read 248741500 bases in 1 contigs\n",
      "[M::mp_idx_build@1.252*1.02] 1943294 blocks\n",
      "[M::mp_idx_build@17.133*0.99] collected syncmers\n",
      "[M::mp_idx_build@17.831*0.99] 70529561 kmer-block pairs\n",
      "[M::mp_idx_print_stat] 1101917 distinct k-mers; mean occ of infrequent k-mers: 64.01; 0 frequent k-mers accounting for 0 occurrences\n",
      "[M::worker_pipeline::18.827*0.99] mapped 40 sequences\n",
      "[M::main] Version: 0.1-r97\n",
      "[M::main] CMD: /domus/h1/msarrias/bin/miniprot/miniprot -t8 fasta_files/GRCh38_with_ce_dup_9_2_after.fa fasta_files/prot_seq9_2.fa\n",
      "[M::main] Real time: 18.873 sec; CPU: 18.711 sec; Peak RSS: 1.599 GB\n",
      "[M::mp_ntseq_read@1.253*1.02] read 248896179 bases in 1 contigs\n",
      "[M::mp_idx_build@1.253*1.02] 1944502 blocks\n",
      "[M::mp_idx_build@17.215*0.99] collected syncmers\n",
      "[M::mp_idx_build@18.015*0.99] 70577503 kmer-block pairs\n",
      "[M::mp_idx_print_stat] 1101937 distinct k-mers; mean occ of infrequent k-mers: 64.05; 0 frequent k-mers accounting for 0 occurrences\n",
      "[M::worker_pipeline::19.006*0.99] mapped 40 sequences\n",
      "[M::main] Version: 0.1-r97\n",
      "[M::main] CMD: /domus/h1/msarrias/bin/miniprot/miniprot -t8 fasta_files/GRCh38_with_ce_dup_10_8_before.fa fasta_files/prot_seq10_8.fa\n",
      "[M::main] Real time: 19.065 sec; CPU: 18.869 sec; Peak RSS: 1.599 GB\n",
      "[M::mp_ntseq_read@1.255*1.02] read 248896179 bases in 1 contigs\n",
      "[M::mp_idx_build@1.255*1.02] 1944502 blocks\n",
      "[M::mp_idx_build@17.609*0.99] collected syncmers\n",
      "[M::mp_idx_build@18.350*0.99] 70577472 kmer-block pairs\n",
      "[M::mp_idx_print_stat] 1101937 distinct k-mers; mean occ of infrequent k-mers: 64.05; 0 frequent k-mers accounting for 0 occurrences\n",
      "[M::worker_pipeline::19.381*0.99] mapped 40 sequences\n",
      "[M::main] Version: 0.1-r97\n",
      "[M::main] CMD: /domus/h1/msarrias/bin/miniprot/miniprot -t8 fasta_files/GRCh38_with_ce_dup_10_8_after.fa fasta_files/prot_seq10_8.fa\n",
      "[M::main] Real time: 19.440 sec; CPU: 19.252 sec; Peak RSS: 1.599 GB\n"
     ]
    }
   ],
   "source": [
    "for setting in settings_list[2:]:\n",
    "    queries = \"fasta_files/prot_seq\" +str(setting[0]) + '_' + str(setting[1]) + \".fa\"\n",
    "    for pos in ['before', 'after']:\n",
    "        hit = \"fasta_files/GRCh38_with_ce_dup_\" + str(setting[0]) + '_' + str(setting[1]) +'_' + pos +  \".fa\"\n",
    "        out_file = 'Miniprot/miniprot_' + str(setting[0]) + '_' + str(setting[1]) + pos + '.txt'\n",
    "        f = open(out_file, \"w\")\n",
    "        subprocess.call([\"/domus/h1/msarrias/bin/miniprot/miniprot\", \"-t8\", hit, queries], stdout=f)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
