{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "553a6db9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from exon_evolution import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d352dfbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "dirct = '/home/msarrias/projects/exon_dups_analysis/code/00-fasta-files/pieris_napi/data/'\n",
    "filename = 'Pieris_napi_brakerProt_rename_agat.gff'\n",
    "db_direct = '/home/msarrias/dbs/pieris_napi.db'\n",
    "\n",
    "## Run only once \n",
    "# create_database(dirct + filename, db_direct)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2f8dec33",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read db using the gffutils library\n",
    "db = gffutils.FeatureDB(db_direct, keep_order=True)\n",
    "annot_dict = get_annotations_dict(db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c47d450b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number</th>\n",
       "      <th>Size total (kb)</th>\n",
       "      <th>Size mean (bp)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>123638</td>\n",
       "      <td>26884.55</td>\n",
       "      <td>217.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <td>123638</td>\n",
       "      <td>26966.50</td>\n",
       "      <td>218.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_UTR</th>\n",
       "      <td>22</td>\n",
       "      <td>62.28</td>\n",
       "      <td>2831.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>32898</td>\n",
       "      <td>220931.03</td>\n",
       "      <td>6715.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron</th>\n",
       "      <td>105783</td>\n",
       "      <td>102665.38</td>\n",
       "      <td>970.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mRNA</th>\n",
       "      <td>17894</td>\n",
       "      <td>129761.45</td>\n",
       "      <td>7251.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_codon</th>\n",
       "      <td>17841</td>\n",
       "      <td>35.68</td>\n",
       "      <td>2.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_codon</th>\n",
       "      <td>17844</td>\n",
       "      <td>35.69</td>\n",
       "      <td>2.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>three_prime_UTR</th>\n",
       "      <td>17</td>\n",
       "      <td>19.63</td>\n",
       "      <td>1154.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <td>16964</td>\n",
       "      <td>114183.29</td>\n",
       "      <td>6730.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>456539</td>\n",
       "      <td>621545.48</td>\n",
       "      <td>26094.23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Number  Size total (kb)  Size mean (bp)\n",
       "CDS              123638         26884.55          217.45\n",
       "exon             123638         26966.50          218.11\n",
       "five_prime_UTR       22            62.28         2831.09\n",
       "gene              32898        220931.03         6715.64\n",
       "intron           105783        102665.38          970.53\n",
       "mRNA              17894        129761.45         7251.67\n",
       "start_codon       17841            35.68            2.00\n",
       "stop_codon        17844            35.69            2.00\n",
       "three_prime_UTR      17            19.63         1154.82\n",
       "transcript        16964        114183.29         6730.92\n",
       "Total            456539        621545.48        26094.23"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stats = pd.DataFrame(basic_stat(annot_dict), index =list(db.featuretypes()) + ['Total'])\n",
    "df_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "67b5daf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_hierarchy_dict = {}\n",
    "for gene in db.features_of_type('gene'):\n",
    "    features = {}\n",
    "    for mRNA_annot in db.children(gene.id, featuretype='mRNA', order_by='start'):\n",
    "        temp_i = []\n",
    "        for child in db.children(mRNA_annot.id, featuretype = ['CDS', 'exon',\n",
    "                                                               'intron', 'five_prime_UTR',\n",
    "                                                               'three_prime_UTR'],\n",
    "                                 order_by='start'):\n",
    "            temp_i += [{'coord': P.open(child.start, child.end),\n",
    "                      'id':child.id, \n",
    "                      'strand': child.strand,\n",
    "                      'type': child.featuretype}]\n",
    "        # sort first by the start and then by the end\n",
    "        temp_j = sorted(temp_i, key = lambda item: (item['coord'].lower, item['coord'].upper))\n",
    "        features[mRNA_annot.id] = temp_j\n",
    "    gene_hierarchy_dict[gene.id] = features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "992c3c63",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check possible overlaps:\n",
    "inters_type_counter = {'exon':0,\n",
    "                       'intron': 0,\n",
    "                       'CDS': 0,\n",
    "                       'five_prime_UTR': 0,\n",
    "                       'three_prime_UTR':0}\n",
    "intersections_counter = {'exon':copy.deepcopy(inters_type_counter),\n",
    "                         'intron': copy.deepcopy(inters_type_counter),\n",
    "                         'CDS': copy.deepcopy(inters_type_counter),\n",
    "                         'five_prime_UTR': copy.deepcopy(inters_type_counter),\n",
    "                         'three_prime_UTR': copy.deepcopy(inters_type_counter)}\n",
    "\n",
    "for gene_id, gene_dict in gene_hierarchy_dict.items():\n",
    "    gene_interval_dict = transcript_interval_dict(gene_dict)\n",
    "    for transcript_id, transcript_dict in gene_interval_dict.items():\n",
    "        overlaping_dict = {}\n",
    "        intervals_list = list(transcript_dict.keys())\n",
    "        for idx, (feat_interv, feature_annot) in enumerate(transcript_dict.items()):\n",
    "            overlaping_dict[feat_interv] = []\n",
    "            if idx != (len(transcript_dict) - 1):\n",
    "                for interval_i in intervals_list[idx+1:]:\n",
    "                    if feat_interv.overlaps(interval_i):\n",
    "                        overlaping_dict[feat_interv].append(interval_i)\n",
    "        for interv, interv_overlap in overlaping_dict.items():\n",
    "            if interv_overlap:\n",
    "                for interval_j in interv_overlap:\n",
    "                    interval_type = transcript_dict[interv]['type']\n",
    "                    next_interval_type = transcript_dict[interval_j]['type']\n",
    "                    if interval_type != next_interval_type:\n",
    "                        intersections_counter[interval_type][next_interval_type] += 1\n",
    "#                         if interval_type=='intron' and next_interval_type == 'exon':\n",
    "#                             print(idx, gene_id, interval_j)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b00dc082",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exon\n",
      "{'exon': 0, 'intron': 3, 'CDS': 33, 'five_prime_UTR': 3, 'three_prime_UTR': 0}\n",
      "   \n",
      "intron\n",
      "{'exon': 14, 'intron': 0, 'CDS': 0, 'five_prime_UTR': 0, 'three_prime_UTR': 0}\n",
      "   \n",
      "CDS\n",
      "{'exon': 6, 'intron': 0, 'CDS': 0, 'five_prime_UTR': 0, 'three_prime_UTR': 0}\n",
      "   \n",
      "five_prime_UTR\n",
      "{'exon': 19, 'intron': 0, 'CDS': 0, 'five_prime_UTR': 0, 'three_prime_UTR': 0}\n",
      "   \n",
      "three_prime_UTR\n",
      "{'exon': 0, 'intron': 0, 'CDS': 0, 'five_prime_UTR': 0, 'three_prime_UTR': 0}\n",
      "   \n"
     ]
    }
   ],
   "source": [
    "for key, value in intersections_counter.items():\n",
    "    print(key)\n",
    "    print(value)\n",
    "    print('   ')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
